src/network/protocol/messages/getdata.zig

```diff
@@ -0,0 +1,165 @@
+        }
+
+        const count = CompactSizeUint.new(self.inventory.len);
+        try count.encodeToWriter(w);
+
+        for (self.inventory) |item| {
+            try w.writeInt(u32, item.type, .little);
+
+            try w.writeAll(&item.hash);
+        }
+    }
+
+    pub fn serialize(self: *const GetdataMessage, allocator: std.mem.Allocator) ![]u8 {
+        const serialized_len = self.hintSerializedLen();
+
+        const ret = try allocator.alloc(u8, serialized_len);
+        errdefer allocator.free(ret);
+
+        try self.serializeToSlice(ret);
+
+        return ret;
+    }
+
+    /// Serialize a message as bytes and write them to the buffer.
+   ///
+   /// buffer.len must be >= than self.hintSerializedLen()
+    pub fn serializeToSlice(self: *const GetdataMessage, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        const writer = fbs.writer();
+        try self.serializeToWriter(writer);
+    }
+
+    pub fn hintSerializedLen(self: *const GetdataMessage) usize {
+        var length: usize = 0;
+
+        // Adding the length of CompactSizeUint for the count
+        const count = CompactSizeUint.new(self.inventory.len);
+        length += count.hint_encoded_len();
+
+        // Adding the length of each inventory item
+        length += self.inventory.len * (4 + 32); // Type (4 bytes) + Hash (32 bytes)
+
+        return length;
+    }
+
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !GetdataMessage {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects reader to have fn 'readInt'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects reader to have fn 'readNoEof'.");
+        }
+
+        const compact_count = try CompactSizeUint.decodeReader(r);
+        const count = compact_count.value();
+
+        const inventory = try allocator.alloc(GetdataMessage.InventoryItem, count);
+
+        for (inventory) |*item| {
+            item.type = try r.readInt(u32, .little);
+            try r.readNoEof(&item.hash);
+        }
+
+        return GetdataMessage{
+            .inventory = inventory,
+        };
+    }
+
+    /// Deserialize bytes into a `GetdataMessage`
+    pub fn deserializeSlice(allocator: std.mem.Allocator, bytes: []const u8) !GetdataMessage {
+        var fbs = std.io.fixedBufferStream(bytes);
+        const reader = fbs.reader();
+        return try GetdataMessage.deserializeReader(allocator, reader);
+    }
+
+
+    pub fn eql(self: *const GetdataMessage, other: *const GetdataMessage) bool {
+        if (self.inventory.len != other.inventory.len) return false;
+
+        for (0..self.inventory.len) |i| {
+            if (self.inventory[i].type != other.inventory[i].type) {
+                return false;
+            }
+            if (!std.mem.eql(u8, &self.inventory[i].hash, &other.inventory[i].hash)) {
+                return false;
+            }
+        }
```

Comment:
Couldn't `std.mem.eql(Inventory, self.inventory, other.inventory)` work

Reply:
@tdelabro  not sure because this is not a simple type.

Reply:
can you impl an eql method on Iventory then

Reply:
yes done